spm_functions.py

```diff
@@ -1,3 +1,119 @@
 # spm_functions.py
 #
-#  This file holds utility functions called by the spm model.
\ No newline at end of file
+#  This file holds utility functions called by the spm model.
+import numpy as np 
+import math
+
+def Half_Cell_Eqlib_Potential(HalfCell,F = 96.48534, T_amb = 298.15, R = 0.0083145):
+    """
+    Help for this function.
+    """
+    # F = 96.48534 #Faraday's number [kC/equivalence]
+    # R = 0.0083145 #Universal gas constant [kJ/mol-K]
+    n_elc = HalfCell.n
+    #T_amb = 273.15 + 25 #[K]
+    T = HalfCell.Temp
+     
+    Delta_G_cell = np.dot(HalfCell.G,HalfCell.nu)
+    Delta_S = np.dot(HalfCell.S,HalfCell.nu)
+
+    U_0_Cell_amb = Delta_G_cell/(n_elc*F)
+    U_0_Cell = U_0_Cell_amb + (T- T_amb)*Delta_S/(n_elc*F)
+    U_Cell = U_0_Cell - R*T/n_elc/F*np.log(np.prod(np.power(HalfCell.X,HalfCell.nu)))
+    return U_Cell
+
+def current_density(i_o,V,U,T,F = 96.48534, Beta = 0.5, R = 0.0083145, n = 1):
+    """
+    The number this returns is absolutly massive, but I cannot find an error in my equation
+    
+    Parameters
+    ----------
+    i_o : Exchange Current Density [mA/cm^2] 
+    V : Electrode potential [V]
+    U : Equilibrium potential [V]
+    T : Temperature of the interface [K]    
+    F : Optional, Faraday's number
+        The default is 96.48534.96.48534 [kC/equivalence]
+    Beta : Optional, The fraction of the total energy that impacts the activation energy of the cathode
+        The default is 0.5
+    R : Optional, Universal gas constant
+        The default is 0.0083145 [kJ/mol-K]
+     n: Optional, number of electrons
+        The default is 1 [equivalence/mol]
+        
+    Returns
+    -------
+    i : current density at the electrode-electrolyte interface [mA/cm^2]
+    """
+    i= i_o*(math.exp((1-Beta)*F*n*(V-U)/(R*T)) - math.exp(-Beta*F*n*(V-U)/(R*T)))
```

Comment:
Two thoughts, from our convo on Tuesday:

1. As we discussed, let's make sure we are intentional about our sign convention. While there is no "correct" convention, I would appreciate sticking with "positive current means positive charge to the electrode,"  just so it is consistent with other work in our group.
2. Typical convention switches `1-Beta` and `Beta` from how you have them, here.
3. When you initialize a reaction, you can store `Beta*n*F/R` and `(1-Beta)*n*F/R` as composite constants, since they are always used together.  Something like `aF_R_fwd = Beta*n*F/R` and `aF_R_ref = (1-Beta)*n*F/R`. (`a` would stand for `alpha`, which is equivalent to `beta*n` or `(1-beta)*n`.  Anyway, this is probably only relevant if we restructure the code where the reaction is a member property of the Half_Cell class.  I don't think we're there yet, but: something to keep in mind.